group/secretsharing/ss.go

```diff
@@ -0,0 +1,138 @@
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/cloudflare/circl/group"
+)
+
+type SecretShare struct {
+	ID    uint64
+	Share group.Scalar
+}
+
+// SecretSharing implements Shamir's secret sharing.
+type SecretSharing interface {
+	Shard(rnd io.Reader, secret group.Scalar) []SecretShare
+	Recover(shares []SecretShare) (secret group.Scalar, err error)
+}
+
+type secretSharing struct {
+	G    group.Group
+	T, N uint64
+}
+
+// New returns a struct implementing SecretSharing interface.
+func New(g group.Group, t, n uint64) (secretSharing, error) {
+	if !(0 < t && t <= n) || g == nil {
+		return secretSharing{}, errors.New("secretsharing: bad parameters")
+	}
+	ss := secretSharing{G: g, T: t, N: n}
+	var _ SecretSharing = ss // checking at compile-time
+	return ss, nil
+}
+
+func (s secretSharing) polyFromSecret(rnd io.Reader, secret group.Scalar) (p polynomial) {
+	p = randomPolynomial(rnd, s.G, s.T)
+	p.coeff[0] = secret.Copy()
+	return
+}
+
+func (s secretSharing) generateShares(poly polynomial) []SecretShare {
+	shares := make([]SecretShare, s.N)
+	x := s.G.NewScalar()
+	for i := range shares {
+		id := uint64(i + 1)
+		x.SetUint64(id)
+		shares[i].ID = id
+		shares[i].Share = poly.evaluate(x)
+	}
+
+	return shares
+}
+
+func (s secretSharing) Shard(rnd io.Reader, secret group.Scalar) []SecretShare {
+	return s.generateShares(s.polyFromSecret(rnd, secret))
+}
+
+func (s secretSharing) Recover(shares []SecretShare) (group.Scalar, error) {
+	if l := len(shares); l <= int(s.T) {
+		return nil, fmt.Errorf("secretsharing: do not met threshold %v with %v shares", s.T, l)
+	} else if l > int(s.N) {
+		return nil, fmt.Errorf("secretsharing: %v shares above max number of shares %v", l, s.N)
+	}
+
+	x := make([]group.Scalar, len(shares))
+	px := make([]group.Scalar, len(shares))
+	for i := range shares {
+		x[i] = s.G.NewScalar()
+		x[i].SetUint64(shares[i].ID)
+		px[i] = shares[i].Share
+	}
+
+	return LagrangeInterpolate(s.G, x, px)
+}
+
+type SharesCommitment = []group.Element
+
+type verifiableSecretSharing struct {
+	s secretSharing
+}
+
+// SecretSharing implements Feldman's secret sharing.
+type VerifiableSecretSharing interface {
+	Shard(rnd io.Reader, secret group.Scalar) ([]SecretShare, SharesCommitment)
+	Recover(shares []SecretShare) (secret group.Scalar, err error)
+	Verify(share SecretShare, coms SharesCommitment) bool
+}
+
+// New returns a struct implementing VerifiableSecretSharing interface.
+func NewVerifiable(g group.Group, t, n uint64) (verifiableSecretSharing, error) {
+	s, err := New(g, t, n)
+	vs := verifiableSecretSharing{s}
+	var _ verifiableSecretSharing = vs // checking at compile-time
+	return vs, err
+}
+
+func (v verifiableSecretSharing) Shard(rnd io.Reader, secret group.Scalar) ([]SecretShare, SharesCommitment) {
+	poly := v.s.polyFromSecret(rnd, secret)
+	shares := v.s.generateShares(poly)
+
+	vecComm := make(SharesCommitment, v.s.T+1)
+	for i, ki := range poly.coeff {
+		vecComm[i] = v.s.G.NewElement()
+		vecComm[i].MulGen(ki)
```

Comment:
We need to add a big fat warning. If our secret is a small scalar, it can easily be recovered from the commitment. E.g., it's easy to see if our secret was 2 from vecComm[0].